### PR TITLE
New registration flow

### PR DESCRIPTION
**Pull request by ppoulin91**
For each subject, register T1 to template, then apply transformation to all `.nii.gz` or `.trk` files

### CODE REVIEW

File: nextflow.config [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+params {
+    help=false
+    input=false
+    template=false
+    output_dir=false
+
+    //**Process control**//
+        processes = false
+
+    Readme_Publish_Dir = "./results/Readme"
```
Comment (arnaudbore) [inline nextflow.config:22]:
```suggestion
    Readme_Publish_Dir = "./results_registration/Readme"
```

File: USAGE [hunk 2]
```diff
@@ -0,0 +1,53 @@
+registration-flow pipeline
+===================
```
Comment (GuillaumeTh) [inline USAGE]: Match the length of your previous line with "=".

File: USAGE [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+OPTIONAL ARGUMENTS (current value)
+
+--output_dir                                Directory where to write the final results.
+                                            By default, will be in "./results"
+
+--processes                                 The number of parallel processes to launch ($cpu_count).
+                                            Only affects the local scheduler.
+
+AVAILABLE PROFILES (using -profile option (e.g. -profile use_cuda,fully_reproducible))
```
Comment (GuillaumeTh) [inline USAGE]: Remove use_cuda and fully_reprocducible and add macos as the example

File: USAGE [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+nextflow run main.nf [OPTIONAL_ARGUMENTS] --input [INPUT] --template [TEMPLATE]
+
+DESCRIPTION
+
+    --input=/path/to/[root]             Root folder containing multiple subjects
+
+                                        [root]
+                                        ├── S1
+                                        |   ├── *t1.nii.gz
```
Comment (arnaudbore) [inline USAGE:17]: Check that you don't have multiple t1 within each folder check number of subject compare with number of t1s

File: main.nf [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+    engine = new groovy.text.SimpleTemplateEngine()
+    template = engine.createTemplate(usage.text).make(bindings)
+
+    print template.toString()
+    return
+}
+
+log.info "Registration pipeline"
+log.info "==================="
```
Comment (GuillaumeTh) [inline main.nf]: Match the length of your previous line with "=".

File: README.md [hunk 6]
```diff
@@ -1,2 +1,17 @@
...
+- [Nextflow](https://www.nextflow.io)
+- [ANTs](http://stnava.github.io/ANTs/)
+  
+If you also want to transform tractograms (.trk):
+- [scilpy](https://github.com/scilus/scilpy)
+
+Usage
+-----
+
+See *USAGE* or run `nextflow run main.nf --help`
```
Comment (arnaudbore) [inline README.md:46]: You should suggest to use registration_flow `release 1.0.0` with  with dockerhub scilus/scilus:1.1.0_registrationflow-1.0.0 :)

File: main.nf [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    .fromFilePairs("$root/**/*.trk",
+                   size:-1,
+                   maxDepth:1,
+                   flat: true) {it.parent.name}
+
+Channel
+    .fromPath("$root/**/*.nii.gz",
+              maxDepth:1)
+    .branch {
+        t1: it =~ "t1.nii.gz"
```
Comment (arnaudbore) [inline main.nf]: does it work if we call the t1 `sub-toto_t1.nii.gz` ?

File: main.nf [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+t1_for_registration
+    .combine(template_for_registration)
+    .set {t1_and_template_for_registration}
+
+
+check_subjects_number.count().set{ number_subj_for_null_check }
+
+number_subj_for_null_check
+.subscribe{a -> if (a == 0)
+    error "Error ~ No subjects found. Please check the naming convention, your --input path or your BIDS folder."}
```
Comment (arnaudbore) [inline main.nf]: Not really a BIDS folder ;)

File: USAGE [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+                                        |   ├── *.nii.gz
+                                        |   └── *.trk
+                                        └── S2
+                                            ├── *t1.nii.gz
+                                            ├── *.nii.gz
+                                            └── *.trk
+
+    --template=/path/to/[template]      Template used for registration in a common space
+
+Files should be compressed Nifti files (.nii.gz) or TRK tractograms (.trk)
```
Comment (arnaudbore) [inline USAGE]: can we be crazy and accept tck with t1 reference ?

Comment (arnaudbore) [reply USAGE]: You could have a specific channel for tck files.

File: README.md (hunk 6 above)
Comment (arnaudbore) [reply README.md:46]: I will push this tag

**Review by arnaudbore — COMMENTED** [review]
Great job, we should be ready to merge and release pretty soon !

File: main.nf [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+if (!params.template){
+    error "Error ~ Please use --template for the template."
+}
+
+root = file(params.input)
+
+Channel.fromPath(file(params.template))
+    .into {template_for_registration; template_for_transformation_nii; template_for_transformation_trk}
```
Comment (GuillaumeTh) [inline main.nf]: If you don't want to have multiple channels for the same template, you can do something like: 

set sid, file(t1) from t1_for_registration
file(template) from template_for_registration.first()

File: main.nf [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+trk_for_transformation = Channel
+    .fromFilePairs("$root/**/*.trk",
+                   size:-1,
+                   maxDepth:1,
+                   flat: true) {it.parent.name}
+
+Channel
```
Comment (GuillaumeTh) [inline main.nf:64]: Not a comment but really cool !

File: main.nf (hunk 8 above)
Comment (GuillaumeTh) [inline main.nf]: Please modify to remove BIDS

File: main.nf [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    echo "[Options]\n" >> readme.txt
+    echo "$list_options" >> readme.txt
+    """
+}
+
+process Register_T1 {
+    cpus params.processes
+
+    input:
+    set sid, file(t1), file(template) from t1_and_template_for_registration
```
Comment (GuillaumeTh) [inline main.nf]: As I said previously you can modify this

File: README.md (hunk 6 above)
Comment (arnaudbore) [reply README.md:46]: Done -> scilus/scilus:1.1.0_registrationflow-1.0.0 exists

File: main.nf [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    .cross(nii_for_transformation)
+    .map { [ it[0][0], it[0][1], it[1][1] ] }
+    .combine(template_for_transformation_nii)
+    .set{nii_and_template_for_transformation}
+
+process Transform_NII {
+    cpus 1
+
+    input:
+    set sid, file(transfo), file(nii), file(template) from nii_and_template_for_transformation
```
Comment (GuillaumeTh) [inline main.nf]: Same here you can have template in an other channel.

File: main.nf [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+    .cross(trk_for_transformation)
+    .map { [ it[0][0], it[0][1], it[1][1] ] }
+    .combine(template_for_transformation_trk)
+    .set{trk_and_template_for_transformation}
+
+process Transform_TRK {
+    cpus 1
+
+    input:
+    set sid, file(transfo), file(trk), file(template) from trk_and_template_for_transformation
```
Comment (GuillaumeTh) [inline main.nf]: Same here you can have template in an other channel.

File: nextflow.config [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+params {
+    help=false
+    input=false
+    template=false
+    output_dir=false
+
+    //**Process control**//
+        processes = false
```
Comment (GuillaumeTh) [inline nextflow.config]: Remove extra space.

**Review by GuillaumeTh — CHANGES_REQUESTED** [review]

File: USAGE (hunk 9 above)
Comment (ppoulin91) [reply USAGE]: @GuillaumeTh  says no... I'll let you guys talk about it

File: nextflow.config [hunk 16]
```diff
@@ -0,0 +1,51 @@
+process {
+    publishDir = {"./results/$sid/$task.process"}
```
Comment (ppoulin91) [inline nextflow.config:2]: Chat with @GuillaumeTh

File: USAGE (hunk 4 above)
Comment (ppoulin91) [reply USAGE:17]: After a chat with @GuillaumeTh , user should be responsible for this, I can push a PR in combine_flows to add my conversion script from TractoFlow

File: README.md [hunk 17]
```diff
@@ -1,2 +1,33 @@
...
+Docker
+------
+If you are on MacOS or Windows, we recommend using the Docker container to run registration-flow
+
+Prebuilt Docker images are available here:
+
+[http://scil.usherbrooke.ca/en/containers_list/](http://scil.usherbrooke.ca/en/containers_list/)
+
+Or use the available Docker image online:
+[scilus/scilus:1.1.0_registrationflow-1.0.0](https://hub.docker.com/layers/scilus/scilus/1.1.0_registrationflow-1.0.0/images/sha256-1726d651667a964f7b0808b8383fac6a92c1c126c45a940da1d7ea5a1693bdb0)
```
Comment (arnaudbore) [inline README.md]: You can just tell them to add this depending on docker or singularity

-with-docker scilus/scilus:1.1.0_registrationflow-1.0.0

or

dl: http://scil.usherbrooke.ca/en/containers_list/scilus-1.1.0_registrationflow-1.0.0.img 

-with-singularity /pathto/scilus-1.1.0_registrationflow-1.0.0.img

File: main.nf (hunk 7 above)
Comment (ppoulin91) [reply main.nf]: Yes: `=~` matches the following regex anywhere in the string

File: USAGE (hunk 4 above)
Comment (ppoulin91) [reply USAGE:17]: I added the validation check :)

**Comment by ppoulin91** [comment]
I have answered all comments here, please review again if everything is ok!

File: README.md [hunk 18]
```diff
@@ -1,2 +1,50 @@
...
+
+FOR DEVELOPERS: The containers repository is available here:
+[containers-TractoFlow](https://github.com/scilus/containers-tractoflow)
+
+
+Docker
+------
+If you are on MacOS or Windows, we recommend using the Docker container to run registration-flow
+
+Simply add the following to yout `nextflow run` command:
```
Comment (arnaudbore) [inline README.md]:
```suggestion
Simply add the following to your `nextflow run` command:
```

File: README.md [hunk 19]
```diff
@@ -1,2 +1,50 @@
...
+`-with-docker scilus/scilus:1.1.0_registrationflow-1.0.0`
+
+
+Prebuilt Docker images are available here:
```
Comment (arnaudbore) [inline README.md]: get rid of these two lines

Comment (arnaudbore) [reply README.md]: now they have all information needed with docker and singularity we don't need to give the link to these singularity images

**Review by arnaudbore — APPROVED** [review]
minor comments LGTM !

**Review by GuillaumeTh — APPROVED** [review]
LGTM ! :beers:

**Pull request merged**